app/src/dex/router/fill_route2.rs

```diff
@@ -0,0 +1,370 @@
+            // constraining.
+
+            let mut constraining_index = None;
+            let mut current_input = Value {
+                amount: input.amount,
+                asset_id: input.asset_id,
+            };
+            tracing::debug!(current_input = ?current_input, "sensing frontier capacity with test amount");
+
+            for (i, position) in frontier.iter().enumerate() {
+                let (unfilled, _new_reserves, output) = position
+                    .phi
+                    .fill(current_input, &position.reserves)
+                    .expect("asset ids should match");
+
+                if unfilled.amount > Amount::zero() {
+                    tracing::debug!(
+                        i,
+                        current_input = ?current_input.amount,
+                        unfilled = ?unfilled.amount,
+                        output = ?output.amount,
+                        "could not completely fill input amount, marking as constraining"
+                    );
+                    // We found a pair that constrains how much we can fill along this frontier.
+                    constraining_index = Some(i);
+                } else {
+                    tracing::debug!(
+                        i,
+                        current_input = ?current_input.amount,
+                        unfilled = ?unfilled.amount,
+                        output = ?output.amount,
+                        "completely filled "
+                    );
+                }
+
+                current_input = output;
+            }
+
+            // Phase 2 (Filling): execute along the path, using the constraint information we sensed above.
+
+            // Special case:
+            // If there was no constraining index, we can completely fill the input amount,
+            // upholding the invariant for loop exits by exhausting the input amount.
+            let Some(constraining_index) = constraining_index else {
+                tracing::debug!("no constraining index found, completely filling input amount");
+
+                let mut current_value = input;
+                trace[0].amount += current_value.amount;
+
+                for (i, position) in frontier.iter_mut().enumerate() {
+                    let (unfilled, new_reserves, output) = position
+                        .phi
+                        .fill(current_value, &position.reserves)
+                        .expect("asset ids should match");
+
+                    assert_eq!(
+                        unfilled.amount,
+                        Amount::zero(),
+                        "unfilled amount for unconstrained frontier should be zero"
+                    );
+
+                    position.reserves = new_reserves;
+                    current_value = output;
+                    trace[i + 1].amount += output.amount;
+                }
+
+                tracing::debug!("filled input amount completely, breaking loop");
```

Comment:
I think we need to update `input.amount` and `output.amount` here.

```rust
            // Update the input and output amounts tracked outside of the loop:
            input.amount = input.amount - current_input.amount;
            output.amount = output.amount + current_output.amount;
 ```

like done further below